src/java.base/share/classes/java/util/HexFormat.java

```diff
@@ -656,13 +661,7 @@ public char toLowHexDigit(int value) {
      */
     public char toHighHexDigit(int value) {
         value = (value >> 4) & 0xf;
-        if (value < 10) {
-            return (char)('0' + value);
-        }
-        if (digitCase == Case.LOWERCASE) {
-            return (char)('a' - 10 + value);
-        }
-        return (char)('A' - 10 + value);
+        return (char) ((value < 10 ? '0' : hexBase) + value);
```

Comment:
These changes seem to be the reason `toHexDigitsLong` consistently sees a slight performance drop. Can any professional engineer explain why the old version, which would be like `(char) (value < 10 ? '0' + value : hexBase + value)`, is slightly faster?

Reply:
Yes, please drop this. Possibly re-attempting it in a future PR, though I'm a bit skeptical of the potential win here.

I did test something like `hexBase + value` in #15591 to a similar effect, so opted for just storing `digitCase`. My best guess as to why are that the JIT is simply better at (or more eager to) eliminating untaken paths than it is at constant fold the field load. Or the branch isn't eliminated and we're seeing the effect of branch prediction. I'd recommend analyzing the asm generated (via e.g. `-prof perfasm`) to understand this in depth